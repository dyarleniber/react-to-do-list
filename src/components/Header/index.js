import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      input: '',
    };
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { input } = this.state;
    const { addTodo } = this.props;
    if (input) {
      addTodo(input);
      this.setState({ input: '' });
    }
  }

  handleInput = (event) => {
    const { value: input } = event.target;
    this.setState({ input });
  }


  render() {
    const { input } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <input type="text" value={input} onChange={this.handleInput} />
        <button type="submit">ADD</button>
      </form>
    );
  }
}

Header.propTypes = {
  addTodo: PropTypes.func.isRequired,
};

export default Header;
