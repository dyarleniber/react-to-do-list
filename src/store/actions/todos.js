import * as types from '../constants/ActionTypes';

export const addTodo = (text) => ({
  type: types.ADD_TODO,
  payload: { text },
});

export const toggleTodo = (id) => ({
  type: types.TOGGLE_TODO,
  payload: { id },
});

export const removeTodo = (id) => ({
  type: types.REMOVE_TODO,
  payload: { id },
});
