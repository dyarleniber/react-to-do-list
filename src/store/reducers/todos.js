import * as types from '../constants/ActionTypes';

const INITIAL_STATE = [];

export default function todos(state = INITIAL_STATE, action) {
  switch (action.type) {
    case types.ADD_TODO:
      return [...state, {
        id: Math.random(),
        text: action.payload.text,
        completed: false,
      }];
    case types.TOGGLE_TODO:
      return state.map((todo) => (
        todo.id === action.payload.id
          ? { ...todo, completed: !todo.completed }
          : todo
      ));
    case types.REMOVE_TODO:
      return state.filter((todo) => todo.id !== action.payload.id);
    default:
      return state;
  }
}
