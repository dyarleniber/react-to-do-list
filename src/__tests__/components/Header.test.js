import React from 'react';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import Header from '../../components/Header';
import { addTodo } from '../../store/actions/todos';

const middlewares = [];
const mockStore = configureStore(middlewares);
const initialState = { };
const store = mockStore(initialState);
const wrapper = mount(
  <Provider store={store}>
    <Header />
  </Provider>,
);

describe('Header component', () => {
  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should change the state', () => {
    const input = 'Input example';
    wrapper.find('[id="textInput"]').first().simulate('change', { target: { value: input } });
    expect(wrapper.find('[id="textInput"]').first().prop('value')).toEqual(input);
  });

  it('should add a new todo', () => {
    const input = 'Input example';
    wrapper.find('[id="textInput"]').first().simulate('change', { target: { value: input } });
    wrapper.find('[id="addForm"]').first().simulate('submit', {
      preventDefault: jest.fn(),
      target: { value: input },
    });
    const actions = store.getActions();
    const text = input.trim();
    const expectedPayload = addTodo(text);
    expect(actions).toEqual([expectedPayload]);
    store.clearActions();
    expect(wrapper.find('[id="textInput"]').first().prop('value')).toEqual('');
  });

  it('should remove white spaces when adding a new todo', () => {
    const rawInput = '   Input example   ';
    const trimmedInput = rawInput.trim();
    wrapper.find('[id="textInput"]').first().simulate('change', { target: { value: rawInput } });
    expect(wrapper.find('[id="textInput"]').first().prop('value')).toEqual(rawInput);
    wrapper.find('[id="addForm"]').first().simulate('submit', {
      preventDefault: jest.fn(),
      target: { value: rawInput },
    });
    const actions = store.getActions();
    const expectedPayload = addTodo(trimmedInput);
    expect(actions).toEqual([expectedPayload]);
    store.clearActions();
    expect(wrapper.find('[id="textInput"]').first().prop('value')).toEqual('');
  });

  it('should do not add white spaces as todos', () => {
    const input = '   ';
    wrapper.find('[id="textInput"]').first().simulate('change', { target: { value: input } });
    expect(wrapper.find('[id="textInput"]').first().prop('value')).toEqual(input);
    wrapper.find('[id="addForm"]').first().simulate('submit', {
      preventDefault: jest.fn(),
      target: { value: input },
    });
    const actions = store.getActions();
    expect(actions).toEqual([]);
    store.clearActions();
    expect(wrapper.find('[id="textInput"]').first().prop('value')).toEqual('');
  });
});
