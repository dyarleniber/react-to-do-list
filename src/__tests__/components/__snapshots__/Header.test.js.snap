// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Header>
    <Container
      className="text-center"
      fluid={false}
    >
      <div
        className="text-center container"
      >
        <Row
          className="align-middle mt-3 mb-4"
          noGutters={false}
        >
          <div
            className="align-middle mt-3 mb-4 row"
          >
            <Col>
              <div
                className="col"
              >
                <img
                  alt="Logo"
                  className="mr-2"
                  src="icon.png"
                />
                <span>
                  <strong>
                    To-Do List
                  </strong>
                </span>
              </div>
            </Col>
          </div>
        </Row>
        <Form
          id="addForm"
          inline={false}
          onSubmit={[Function]}
        >
          <form
            className=""
            id="addForm"
            onSubmit={[Function]}
          >
            <FormRow
              className="mb-3"
            >
              <div
                className="mb-3 form-row"
              >
                <FormGroup
                  as={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "displayName": "Col",
                      "render": [Function],
                    }
                  }
                  md={
                    Object {
                      "offset": 3,
                      "span": 6,
                    }
                  }
                >
                  <Col
                    className="form-group"
                    md={
                      Object {
                        "offset": 3,
                        "span": 6,
                      }
                    }
                  >
                    <div
                      className="form-group col-md-6 offset-md-3"
                    >
                      <FormControl
                        autoComplete="off"
                        id="textInput"
                        maxLength="100"
                        onChange={[Function]}
                        placeholder="What needs to be done?"
                        type="text"
                        value=""
                      >
                        <input
                          autoComplete="off"
                          className="form-control"
                          id="textInput"
                          maxLength="100"
                          onChange={[Function]}
                          placeholder="What needs to be done?"
                          type="text"
                          value=""
                        />
                      </FormControl>
                    </div>
                  </Col>
                </FormGroup>
                <Col
                  md={
                    Object {
                      "offset": 3,
                      "span": 6,
                    }
                  }
                >
                  <div
                    className="col-md-6 offset-md-3"
                  >
                    <Button
                      active={false}
                      className="float-right"
                      disabled={false}
                      id="addButton"
                      size="sm"
                      type="submit"
                      variant="primary"
                    >
                      <button
                        className="float-right btn btn-primary btn-sm"
                        disabled={false}
                        id="addButton"
                        type="submit"
                      >
                        <small>
                          +
                           
                          ADD
                        </small>
                      </button>
                    </Button>
                  </div>
                </Col>
              </div>
            </FormRow>
          </form>
        </Form>
      </div>
    </Container>
  </Header>
</Provider>
`;
